domain/cloudimagemetadata/state/state.go

```diff
@@ -0,0 +1,354 @@
+INSERT INTO cloud_image_metadata (*)
+VALUES ($inputMetadata.*)`, inputMetadata{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		toInsert := make([]inputMetadata, 0, len(values))
+		for _, m := range values {
+			var out metadataUUID
+			err := tx.Query(ctx, checkExistMetadataStmt, m).Get(&out)
+			if dberrors.IsErrNotFound(err) {
+				// Not found => insert
+				toInsert = append(toInsert, m)
+			} else if err == nil {
+				// found => Update
+				m.UUID = out.UUID
+				err = tx.Query(ctx, updateMetadataStmt, m).Run()
+				if err != nil {
+					return errors.Trace(err)
+				}
+			} else {
+				return errors.Trace(err)
+			}
+		}
+
+		// Insert images metadata
+		if len(toInsert) == 0 {
+			return nil
+		}
+		return errors.Trace(tx.Query(ctx, insertMetadataStmt, toInsert).Run())
+	}))
+}
+
+// DeleteMetadataWithImageID deletes all metadata associated with the given image ID from the database.
+func (s *State) DeleteMetadataWithImageID(ctx context.Context, imageID string) error {
+	db, err := s.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	deleteMetadataStmt, err := sqlair.Prepare(`
+DELETE FROM cloud_image_metadata 
+WHERE image_id = $metadataImageID.image_id`, metadataImageID{})
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	return errors.Trace(db.Txn(ctx, func(ctx context.Context, tx *sqlair.TX) error {
+		err := tx.Query(ctx, deleteMetadataStmt, metadataImageID{imageID}).Run()
+		return err
+	}))
+}
+
+// FindMetadata retrieves cloud image metadata from the database based on specified filter criteria.
+// It constructs a dynamic SQL query using the supplied criteria and executes it to fetch matching records.
+// Returns [cloudmetadataerrors.NotFound] if none are found with this criteria.
+func (s *State) FindMetadata(ctx context.Context, criteria cloudimagemetadata.MetadataFilter) ([]cloudimagemetadata.Metadata, error) {
+	db, err := s.DB()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	filter := metadataFilter{
+		Region:          criteria.Region,
+		Versions:        criteria.Versions,
+		Arches:          criteria.Arches,
+		Stream:          criteria.Stream,
+		VirtType:        criteria.VirtType,
+		RootStorageType: criteria.RootStorageType,
+	}
+
+	// clauses will collect all required clauses to build the final WHERE ... AND ... clause
+	clauses := []string{
+		// Ignores expired metadata
+		fmt.Sprintf(`cloud_image_metadata.created_at >=  '%s'`, s.clock.Now().Add(-ExpirationDelay).Format(time.DateTime)),
+	}
+
+	// declareEqualsClause is an helper function to add a sqlair clause with the format cloud_image_metadata.<field> = $metadataFilter.<field>,
+	// only if the corresponding field isn't empty
+	hasFilter := false
+	declareEqualsClause := func(field, value string) {
+		if value != "" {
+			hasFilter = true
+			clauses = append(clauses, fmt.Sprintf(`cloud_image_metadata.%s = $metadataFilter.%s`, field, field))
```

Comment:
We could have used a map here.

Reply:
I had also this idea, but it seems a bit overkill, and not aligned with the intent.

We have `0..1` filter, not `0..n`.

So I discard it, unless you disagree.

Reply:
Isn't this clearer?
```
if filter.Region != "" {
    clauses = append(clauses, `cloud_image_metadata.region = $metadataFilter.region`)
}
if filter.Stream != "" {
    clauses = append(clauses, `cloud_image_metadata.stream = $metadataFilter.stream`)
}
if filter.VirtType != "" {
    clauses = append(clauses, `cloud_image_metadata.virt_type = $metadataFilter.virt_type`)
}
if filter.RootStorageType != "" {
    clauses = append(clauses, `cloud_image_metadata.root_storage_type = $metadataFilter.root_storage_type`)
}
```